### DISCUSSION TITLE
Framework for rust-backed custom expressions

### DISCUSSION THREAD

**Discussion by mjclarke94** [Eventual-Inc/Daft#3439] [Ideas]
One of the nice features polars offers is the ability to write [custom plugins](https://docs.pola.rs/user-guide/plugins/your-first-polars-plugin/) backed by rust. It isn't clear from the documentation whether this is possible (or indeed, encouraged!) in Daft. 

Whilst UDF's are great, it's clear that when performance is critical, writing expressions in rust can lead to huge performance improvements (see performance comparison for the implementation of MinHash as an example: https://github.com/Eventual-Inc/Daft/pull/2431). 

Something analogous to `pyo3-polars` would be fantastic in situations where you have really heavy/domain specific operations to scale and are happy to put the time in creating a custom expression.

**Comment by jaychia** [comment]
@colin-ho @samster25  thoughts?

This sounds really reasonable, and we did already do a big refactor to try and make our codebase more modular/extensible such that we can start supporting external plugins for expressions.
